lib/elixir/test/elixir/kernel/typespec_test.exs

```diff
@@ -642,4 +642,54 @@ defmodule Kernel.TypespecTest do
     assert Kernel.Typespec.beam_types(Unknown) == nil
     assert Kernel.Typespec.beam_specs(Unknown) == nil
   end
+
+  defmodule Sample do
+    @callback first(integer) :: integer
+
+    @callback foo(atom(), binary) :: binary
+
+    @callback bar(External.hello, my_var :: binary) :: binary
+
+    @callback guarded(my_var) :: my_var when my_var: binary
+
+    @callback orr(atom | integer) :: atom
+
+    @callback literal(123, {atom}, :atom, [integer], true) :: atom
+
+    @macrocallback last(integer) :: Macro.t
+  end
+
+  test :callbacks do
+    assert Sample.behaviour_info(:callbacks) == [first: 1, guarded: 1, "MACRO-last": 2, literal: 5, orr: 1, foo: 2, bar: 2]
+  end
+
+  test :specs do
+    assert length(Keyword.get_values(Sample.module_info[:attributes], :callback)) == 7
+  end
+
+  test :default_is_not_supported do
+    assert_raise ArgumentError, fn ->
+      defmodule WithDefault do
+        @callback hello(num \\ 0 :: integer) :: integer
```

Comment:
Let's use `@spec` in some of those assertions just to prove the point they are not allowed anywhere. :)